includes/data-stores/class-wc-admin-reports-customers-data-store.php

```diff
@@ -0,0 +1,523 @@
+			'page'     => 1,
+			'order'    => 'DESC',
+			'orderby'  => 'date_registered',
+			'fields'   => '*',
+		);
+		$query_args = wp_parse_args( $query_args, $defaults );
+
+		$cache_key = $this->get_cache_key( $query_args );
+		$data      = wp_cache_get( $cache_key, $this->cache_group );
+
+		if ( false === $data ) {
+			$data = (object) array(
+				'data'    => array(),
+				'total'   => 0,
+				'pages'   => 0,
+				'page_no' => 0,
+			);
+
+			$selections       = $this->selected_columns( $query_args );
+			$sql_query_params = $this->get_sql_query_params( $query_args );
+
+			$db_records_count = (int) $wpdb->get_var(
+				"SELECT COUNT(*)
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+				"
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			$total_pages = (int) ceil( $db_records_count / $sql_query_params['per_page'] );
+			if ( $query_args['page'] < 1 || $query_args['page'] > $total_pages ) {
+				return $data;
+			}
+
+			$customer_data = $wpdb->get_results(
+				"SELECT
+						{$selections}
+					FROM
+						{$table_name}
+						{$sql_query_params['from_clause']}
+					WHERE
+						1=1
+						{$sql_query_params['where_time_clause']}
+						{$sql_query_params['where_clause']}
+					GROUP BY
+						customer_id
+					ORDER BY
+						{$sql_query_params['order_by_clause']}
+					{$sql_query_params['limit']}
+					",
+				ARRAY_A
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			if ( null === $customer_data ) {
+				return $data;
+			}
+
+			$customer_data = array_map( array( $this, 'cast_numbers' ), $customer_data );
+			$data          = (object) array(
+				'data'    => $customer_data,
+				'total'   => $db_records_count,
+				'pages'   => $total_pages,
+				'page_no' => (int) $query_args['page'],
+			);
+
+			wp_cache_set( $cache_key, $data, $this->cache_group );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Add customer information to the lookup table when orders are created or modified.
+	 *
+	 * @param int $post_id Post ID.
+	 */
+	public static function sync_order( $post_id ) {
+		if ( 'shop_order' !== get_post_type( $post_id ) ) {
+			return;
+		}
+
+		$order = wc_get_order( $post_id );
+		if ( ! $order ) {
+			return;
+		}
+
+		$customer_id = $order->get_customer_id();
+		if ( 0 === $customer_id ) {
+			return;
+		}
+
+		self::update_registered_customer( $customer_id );
+	}
+
+	/**
+	 * Updates a guest (no user_id) customer data with new order data.
+	 *
+	 * @param WC_Order $order Order to update guest customer data with.
+	 * @return int|false The number of rows affected, or false on error.
+	 */
+	public function update_guest_customer_by_order( $order ) {
+		global $wpdb;
+
+		$email = $order->get_billing_email( 'edit' );
+
+		if ( empty( $email ) ) {
+			return true;
+		}
+
+		$existing_guest = $this->get_guest_by_email( $email );
+
+		if ( $existing_guest ) {
```

Comment:
If we make changes to customer stats below, we might be able to clean up this logic a bit by using `$wpdb->replace()` and appending the customer ID.

[Example of this](https://github.com/woocommerce/wc-admin/blob/20fad383849bd33094f5893f3238f1bd46f5fb6d/includes/data-stores/class-wc-admin-reports-customers-data-store.php#L479).

Reply:
Yep, we could just rework [`update_registered_customer()`](https://github.com/woocommerce/wc-admin/pull/1144/files/1399528579267b959af4bbefec04e57a200c4f20#diff-061c71b3c8238141695a290a5becc8baR466) instead of having multiple functions.